### DISCUSSION TITLE
Is there any way to type hint a non-None hashable?

### DISCUSSION THREAD

**Discussion by JanEricNitschke** [microsoft/pyright#8265] [General]
Specifically for a function like:

```
def func(a: Hashable, b: Hashable) -> bool:
    if a is None and b is None:
        raise ValueError("Both arguments cannot be None")
    # Do something that requires them to be hashable
```

Again specifically in the usecase of pandas stubs where `df.drop()` raises this error `ValueError: Need to specify at least one of 'labels', 'index' or 'columns'` which i would like to encode in the overloaded signatures.

**Comment by JanEricNitschke** [comment]
The only idea i have at the moment would be to add an overload of

`def func(a: None, b: None) -> Never`. 

But that would probably only handle the case where you want to assign the return value and not where it is used for side effects.

**Reply by erictraut** [reply]
A `Never` return type doesn't mean that a call is invalid, calling such an overload will not generate an error.

There has been some talk about adding the ability to mark specific overloads as an "error" (often in the context of pandas), but these discussions have never converged on a solution. You could mark the overload as `@deprecated`, but that's not quite the same thing.

**Reply by JanEricNitschke** [reply]
Got it, thanks! I slightly messed up the example above (full thing is again from pandas-stubs) and missed a detail. Namely that these all have defaults.

```
def func(a: Hashable = None, b: Hashable = None) -> bool:
    if a is None and b is None:
        raise ValueError("Both arguments cannot be None")
    # Do something that requires them to be hashable
```

My current approach is to not allow the cases where everything has the default by just not allowing the overload where everything has a default. But that does leave people open to doing `func(None, None)`.

I guess i could add the explicit overload with the double None and mark it as deprecated. I could also additional make the return type `Never` so that there is at least an issue when someone does `a = func(None, None)` or is that two indirect. The error message would probably not be helpful.

**Reply by erictraut** [reply]
Yeah, that approach sounds reasonable. However, there won't be an error message for `a = func(None, None)` because `Never` is assignable to all types.

**Reply by JanEricNitschke** [reply]
Ok, then ill check with the stubs maintainer what they think about that part.

I guess the optimal solution would be if there was an explicit way to exclude a type like `NonNoneHasheable = Exclude[None, Hasheable]` which would also help with the `str` is a `Sequence[str]` think, but i assume there are again some intricate issue somewhere with that and its not on the horizon.
